pyproject.toml

```diff
@@ -73,6 +73,7 @@ dependencies = [
   "python-docx",
   "langdetect",  # for PDF conversions
   "tika",  # Apache Tika (text & metadata extractor)
+  "PyMuPDF>=1.18.16" ,  # PDF text extraction
```

Comment:
This creates a conflict with [Apache-2.0 license](https://github.com/deepset-ai/haystack/blob/main/LICENSE):
https://mupdf.com/licensing/index.html

Reply:
That's an interesting point. I'm uncertain if this usage violates any of the restrictions of the Affero-GPL v3: https://www.gnu.org/licenses/agpl-3.0.fr.html

Reply:
A lawyer perspective on AGPLv3:

https://writing.kemitchell.com/2021/01/24/Reading-AGPL

> The Condition
> Recall that the condition had two prongs: modifying the program and supporting network interaction. If you’re thinking that was an odd way to write it, you’d be right. Especially if we know a little bit about why GPL was patched to make AGPL in the first place. Which we can glean from the preamble:
> 
> A secondary benefit of defending all users’ freedom is that improvements made in alternate versions of the program, if they receive widespread use, become available for other developers to incorporate. Many developers of free software are heartened and encouraged by the resulting cooperation. However, in the case of software used on network servers, this result may fail to come about. The GNU General Public License permits making a modified version and letting the public access it on a server without ever releasing its source code to the public.
> 
> The GNU Affero General Public License is designed specifically to ensure that, in such cases, the modified source code becomes available to the community. It requires the operator of a network server to provide the source code of the modified version running there to the users of that server. Therefore, public use of a modified version, on a publicly accessible server, gives the public access to the source code of the modified version.

> In the preamble, we learn that AGPL puts new requirements on operators of network servers. But we just read the network copyleft paragraph of section 13, and its condition didnt’t mention operating a network service. It kicks in when we modify the program and our modified version supports being used to operate a network service.

Reply:
@maxjeblick how so? We don't modify the original source code nor we distribute it directly. It would be different if this PR contained source modules from PyMuPDF, then yes Haystack as a whole would need to be distributed under AGPLv3. 

I might still see how somebody might want to avoid bringing in AGPL software, but if that's the concern maybe we can make it an optional dep under a setup extras?

Reply:
To my understanding, including PyMuPDF as a dependency would convert haystack's license to AGPLv3 (as haystack would include some of PyMuPDF's functionality).

Reply:
My perspective is similar to [Heather Meeker](https://www.synopsys.com/blogs/software-security/agpl-affero-gpl-3/), lawyer: 

> Enduring questions
> Unfortunately, the language of the additional network use condition in AGPL leaves some questions unanswered. The most common question is how far “down the stack” the condition reaches. Suppose you are using AGPL software as an operating system or language component to serve up an application. The user interacts with the application, but not with this infrastructure element. The copyleft condition says you must offer source code to “users interacting with it [i.e. the Program] remotely through a computer network.” That seems to mean that infrastructure elements will never trip this condition, because any user interaction would be too indirect.

> Plato’s cave
> Not surprisingly, AGPL3 has been slower to propagate than GPL3; if GPL3 has gained acceptance slowly, AGPL3 has been tagging along behind like a younger sibling. But the license is not as scary as it seems. The network interaction provision only triggers source code offer requirements when you modify the Program. In fact, most open source software is used without modification from community versions. After all, most businesses use open source because it is already tried and tested, free of charge, and supported by the community. Modifying it only undercuts those advantages. So, particularly with killer apps like MongoDB, many companies have become cautiously comfortable approving AGPL software for SaaS use. They still worry—and probably rightly so—that modifications could be made to the code after approval, and they lack internal controls to revisit license compliance when the modifications occur.

AGPLv3 is quite misunderstood. And even some people which uses AGPLv3 as their open-source licensing model, sometime uses it trying to create a freeware. Which is not the case, AGPLv3 is not a freeware license.

Reply:
We can add to the list of open-source projects which are under AGPLv3 and don't impose to anybody who makes usage of them to convert it's licensing to AGPLv3:

- MongoDB
- Grafana
- Minio

Imagine that each deployment which used that stack would need to change it's licensing. Many corporations would need to turn their product's license into AGPLv3.
You may say that PyMuPDF is a module, different from the others, which are not. But the license applies to the same direction: source code. That source code produces a result, in the case of PyMuPDF, PDF handling. When we are using the outputs of the module in a public distribution repository (PyPi), and our code clearly shows how it was used 
(good faith demonstration principle), without changing the source code, we are not breaking AGPLv3 in any point. Nor obligating Haystack to convert it's licensing, as the _source-code_ was not incorporated and changed.

So, there is no:
- **distribution** -> The package is not distributed by Haystack, but by PyPi, a public repository
- **remote network interaction** -> We are not exposing PyMuPDF directly over the internet to users, and this will probably never be the case to Haystack users. Haystack being a framework, it will be used to build _backend infrastructure_.
- **corresponding source** -> We are not providing any custom installation scripts, methods even without changing the source code. The package is installed from a public repository, PyPi.
- **source code change** -> We have _not changed_ any of the original _PyMuPDF original source code_, and this _repository_ is the _source of truth_ of such statement. Haystack is open source, and the usage of PyMuPDF is open to any analysis by any-third party.

Of course, there are cases where some companies try to create additional restrictions, but international law doesn't accept such “estoppel” technique.

On the other side, a violation of AGPLv3 would occur if someone:
- creates a SaaS which has as only reason the PDF handling, even without changing the code.
- change the source code and keep it private
- create custom installation mechanisms for the package

Reply:
In any case, I'm refactoring this PR to support both xpdf and PyMuPDF. The default installation will use xpdf, and each user will be able to choose to install PyMuPDF as an extra: `[pdf]`. I'll reopen it when finished.

Reply:
@danielbichuetti I got further confirmation we are good to go with `PyMuPDF` so I would actually ask you to keep it like this if the only concern is about the license.

Reply:
Thanks a lot for the discussion, I'm not an expert w.r.t. to licensing, and this is certainly helpful.
I also saw that this repo uses https://fossa.com/ for license scanning :)

Reply:
@masci Thank you for the approval.

I was pretty sure about my perspective because after what @maxjeblick mentioned, I took some hours reading and investigating court cases in the US and BR regarding open-source licensing, dual licensing and multi licensing. Furthermore, I asked for advice from one of our co-founders, which is a lawyer authorized to work both in BR (OAB) and the US (BAR). Going further, I did a free consultancy with @rolantannous, as he is a commercial mediator.

All the replies are condensed in the previous posts in an easy-to-understand explanation.

I was making changes to support both scenarios, as this conversation created a small concern about common user understanding of licensing. I'll revert them and reopen the PR.